primo2/inference/dynamic.py

```diff
@@ -0,0 +1,305 @@
+            The variable/node.
+
+        Returns
+        -------
+        dict
+            A dictionary, dict-keys are the values of the variable/node and 
+            dict-values are the probabilities of each value.
+        """
+        pass
+
+    @abc.abstractmethod
+    def marginals(self, variables):
+        """Get the marginals of a variable. (abstract)
+
+        Parameters
+        ----------
+        variables : [String,]
+            List containing the names of the variables whose joint prior
+            or posterior marginals are desired.   
+                
+        Returns
+        -------
+        ?
+            The marginals.
+        """
+        pass
+
+    def set_evidence(self, evidence=None, **kwargs):
+        """Set evidence for the current time slice, without proceeding in time.
+
+        Calls the method `_set_evidence`, which does the actual work.
+        
+        Parameters
+        ----------
+        evidence : dict, optional
+            Dictionary containing the given evidence. The keys represent the
+            variable names of the given evidence and the values can either
+            be simple strings of the given values, or a np.array specifying
+            the strength of the evidence for each outcome, e.g.: For the binary
+            evidence node E `{"E": "True"}` is equivalent to
+            `{"E": np.array([1.0, 0.0])}`.
+        """
+        r = self._set_evidence(evidence)
+        self._evidence = True
+        return r
+
+    def unroll(self, evidence=None, **kwargs):
+        """Unroll the network, and proceed in time.
+
+        If no evidence has been set (via `set_evidence`) or is specified in the
+        arguments it proceeds without evidence. If evidence has been set
+        previously and no evidence is specified in the arguments it uses this
+        pre-specified evidence. If evidence is set in the arguments, this new
+        evidence is used.
+
+        Calls the method `_unroll`, which does the actual work.
+
+        evidence : dict, optional
+            Dictionary containing the given evidence. The keys represent the
+            variable names of the given evidence and the values can either
+            be simple strings of the given values, or a np.array specifying
+            the strength of the evidence for each outcome, e.g.: For the binary
+            evidence node E `{"E": "True"}` is equivalent to
+            `{"E": np.array([1.0, 0.0])}`.
+        **kwargs
+            Takes the same keyword arguments as `set_evidence` and
+            `_set_evidence`.
+        """
+        if not self._evidence or evidence is not None:
+            self._set_evidence(evidence, **kwargs)
+        r = self._unroll()
+        self._evidence = False
+        return r
+
+    @abc.abstractmethod
+    def _set_evidence(self, evidence=None, **kwargs):
+        """Set evidence for the network (doing the actual work). (abstract)
+
+        See `set_evidence`.
+        """
```

Comment:
We could add something like "Needs to be implemented by subclasses" just to make these docstrings more accurate by themselves but this feels redundant with the decorator.